pennylane/capture/meta_type.py

```diff
@@ -0,0 +1,163 @@
+"""
+Defines a metaclass for automatic integration of any ``Operator`` with plxpr program capture.
+
+See ``explanations.md`` for technical explanations of how this works.
+"""
+
+import abc
+from functools import lru_cache
+
+import pennylane as qml
+
+from .switches import plxpr_enabled
+
+has_jax = True
+try:
+    import jax
+except ImportError:
+    has_jax = False
+
+
+@lru_cache  # constrcut the first time lazily
+def _get_abstract_operator():
+    """Create an AbstractOperator once in a way protected from lack of a jax install."""
+    if not has_jax:
+        raise ImportError("Jax is required for plxpr.")
+
+    class AbstractOperator(jax.core.AbstractValue):
+        """An operator captured into plxpr."""
+
+        # pylint: disable=missing-function-docstring
+        def at_least_vspace(self):
+            # TODO: investigate the proper definition of this method
+            raise NotImplementedError
+
+        # pylint: disable=missing-function-docstring
+        def join(self, other):
+            # TODO: investigate the proper definition of this method
+            raise NotImplementedError
+
+        # pylint: disable=missing-function-docstring
+        def update(self, **kwargs):
+            # TODO: investigate the proper definition of this method
+            raise NotImplementedError
+
+        def __eq__(self, other):
+            return isinstance(other, AbstractOperator)
+
+        def __hash__(self):
+            return hash("AbstractOperator")
+
+        @staticmethod
+        def _matmul(*args):
+            return qml.prod(*args)
+
+        @staticmethod
+        def _mul(a, b):
+            return qml.s_prod(b, a)
+
+        @staticmethod
+        def _rmul(a, b):
+            return qml.s_prod(b, a)
+
+        @staticmethod
+        def _add(a, b):
+            return qml.sum(a, b)
+
+        @staticmethod
+        def _pow(a, b):
+            return qml.pow(a, b)
+
+    jax.core.raise_to_shaped_mappings[AbstractOperator] = lambda aval, _: aval
+
+    return AbstractOperator
+
+
+class MetaOperator(abc.ABCMeta):
+    """A metatype that:
+
+    * automatically registers a jax primitive to ``cls._primitive``
+    * Dispatches class creation to ``cls._primitive.bind`` instead of normal class
+    creation when the primitive is defined and plxpr capture is enabled.
+
+    See ``pennylane/capture/explanations.md`` for more detailed information on how this technically
+    works.
+
+    """
+
+    def __init__(cls, *args, **kwargs):
+
+        super().__init__(cls, args, kwargs)
+
+        # Called when constructing a new class that has this metaclass.
+        # Similar to __init_subclass__ , this allows us to run this code
+        # every time we define a new class
+
+        if not has_jax:
+            cls._primitive = None
+            return
+
+        cls._primitive = jax.core.Primitive(cls.__name__)
+
+        @cls._primitive.def_impl
+        def _(*args, **kwargs):
+            if "n_wires" not in kwargs:
```

Comment:
I add it in during `_primitive_bind_call`.   Because we are currently unpacking and repacking all the individual wires, we need to know how to separate out the normal arguments from the wire arguments.

Reply:
`type` is a python built-in.  You can think of as the default metaclass or the parent of all python types.  For example:

```
>>> isinstance(int, type)
True
>>> isinstance(qml.X, type)
True
```

So by calling these methods on `type`, I'm essentially restoring default python behavior.

Reply:
Thanks for the explanation